### DISCUSSION TITLE
Hybrid ABM DEVS with MESA

### DISCUSSION THREAD

**Discussion by quaquel** [mesa/mesa#2032] [Show and tell]
I have mentioned in various discussions the potential for hybrid ABM-Discrete Event scheduling. After finishing the new cell grid stuff and being a bit stuck on data collection, I decided to give it a go. The code uses ideas from the current devs scheduler and [pydsol](https://github.com/averbraeck/pydsol-core/tree/main), which is a python port of the DSOL java library developed over the past 20 years by colleagues. 

My draft code is [here](https://github.com/quaquel/mesa/tree/devs/mesa/experimental/devs). The key classes are a new Simulator class (akin to a numerical integrator for ODEs, theoretically rooted in the work of [Zeigler](https://www.sciencedirect.com/book/9780128133705/theory-of-modeling-and-simulation)), the EventList, and SimEvent. 

I also added so far 1 example: [wolfsheep](https://github.com/quaquel/mesa/blob/devs/mesa/experimental/devs/examples/wolf_sheep.py). The key thing is that I use ABM-style scheduling on the wolf and sheep.

```python
    def step(self):
        self.get_agents_of_type(Sheep).shuffle(inplace=True).do("step")
        self.get_agents_of_type(Wolf).shuffle(inplace=True).do("step")
        self.simulator.schedule_event_relative(self.step, 1)
```

While the regrowth of the grass patches is handled through event scheduling. So, the grass patches are no longer activated each step to reduce a counter. Rather, I schedule events for when they are fully regrown.

```python
    @fully_grown.setter
    def fully_grown(self, value):
        self._fully_grown = value

        if value == False:
            self.model.simulator.schedule_event_relative(self.set_fully_grown, self.grass_regrowth_time)
``` 

It is just a first rough example. I want to do Epstein's civil violence next, with jail time being handled through event scheduling. A key thing, however, is that wolf sheep on 100 ticks already has about half the runtime of the pure ABM implementation. 

**added 1**: Currently, the core design uses [next event time progression](https://en.wikipedia.org/wiki/Discrete-event_simulation). To get AMB style behavior (_i.e._, incremental time progression), `model.step` schedules itself with a time increment of 1. A slightly more elaborate implementation might make this the default behavior for the `ABMSimulator`, while still allowing event scheduling. So, ABMSimulator schedules/always contains a `model.step` for each discrete time tick. From a user point of view, this might be more convenient.

**added 2**: I am slowly continuing to update the code. Incremental time progression has now been integrated, so in `model.step` it is no longer needed to schedule the next call to `model.step`. Moreover, I have given `model.step` events a higher priority so they always go first in a tick.

**added 3**: PR:
- https://github.com/projectmesa/mesa/pull/2066

**Comment by rht** [comment]
Existing discussions:
1. Initial proposal https://github.com/projectmesa/mesa/issues/1893
1. Motivation https://github.com/projectmesa/mesa/discussions/1997#discussioncomment-8370650
1. Performance advantage of DEVS over ABM scheduling https://github.com/projectmesa/mesa/discussions/2027#discussioncomment-8395278
1. Initial implementation (merged) https://github.com/projectmesa/mesa/pull/1890

Will add more to the list when a new relevant / related discussion appears.
Note: I consent for the maintainers to edit this post for updating the list.

**Comment by rht** [comment]
Replying to https://github.com/projectmesa/mesa/discussions/1997#discussioncomment-8370650

> Practically, it would involve a bit more than making a more comprehensive scheduler. Rather, it means rebuilding MESA on top of an EventTree. Practically, this might sound bigger than it. Zeigler makes a distinction between a model and a simulator. ...

Does DSOL/pydsol use the concept of EventTree? The term EventTree needs an elaboration itself, and I'm not sure if it is discussed in [Zeigler 2018](https://www.sciencedirect.com/book/9780128133705/theory-of-modeling-and-simulation) either.

**Reply by quaquel** [reply]
pydsol uses "eventlist" instead of "event tree". So, I switched to using that as well.  My original term in #1997 was simply incorrect. 

Also, note that the formalism I use here is called event scheduling (see also [Zeigler](https://www.sciencedirect.com/science/article/pii/B9780128133705000365#is0030)]

**Comment by Corvince** [comment]
Thank you @quaquel for putting this together!! Lots of interesting things in there and I am looking forward to diving into it

**Comment by EwoutH** [comment]
Some previous discussion, for reference:
- https://github.com/projectmesa/mesa/pull/1890
- https://github.com/projectmesa/mesa/issues/1893

It would be interesting how many goals from #1893 we can integrate.

**Reply by quaquel** [reply]
some quick thoughts

1. My implementation can cancel events, so this offers some basis upon which to build interruption. The key is that you need to know exactly which event to cancel. 
2. My implementation allows you to schedule any callable, not just agent.step. So system level events are included by default. It might be possible/desirable to refine this. I currently only accept a callable, but you might make "AgentEvents" where you have the object (i.e., agent) and the method you want to schedule. This makes it easier to remove events belonging to a specific agent.
3. My implementation does not use weakrefs in contrast to DiscreteEventScheduler. So agent removal if they have still events scheduled won't work. Fixing this is straightforward and on my todo list.
4. Resumption logic seems more of an agent issue not an event scheduling issue.

**Reply by rht** [reply]
The list has been merged into https://github.com/projectmesa/mesa/discussions/2032#discussioncomment-8436111.

**Comment by wang-boyu** [comment]
I briefly discussed with my PhD advisor Andrew, and he likes the idea of making Mesa a hybrid simulation framework. Recently MASON (a JAVA ABM framework) has a discrete event simulation extension named MASON DES, aiming to provide such hybrid simulation modes ([link to paper](https://ieeexplore.ieee.org/abstract/document/10155378)). There is a repository containing three example DES models ([link to repo](https://github.com/eclab/DES-Supply-Chain-demo)) but I couldn't find the source code for the DES extension. Perhaps it's not yet released.

For Mesa, I'm wondering if we'd like the DES part to be a separate extension (like MASON DES to MASON), or have it integrated within Mesa.

I just realized that there might be a slight difference between the goals. Do we want (1) just a discrete event scheduler, or (2) the capabilities to develop discrete event models in Mesa? For (1), we are still doing ABMs, but we can develop in DES-style; whereas (2) refers to a full-blown framework to do DES models (e.g., a queuing system).

**Reply by quaquel** [reply]
DES has three basic paradigms: Event Scheduling, Activity Scanning, and Process interaction. In the implementation provided here, I used Event Scheduling because, based on previous work from colleagues, I knew this was the one that was most easily aligned with ABMs. 

As you can see in [simulator.py](https://github.com/quaquel/mesa/blob/devs/mesa/experimental/devs/simulator.py), I currently implement both an ABMSimulator and a DEVSSimulator. The ABM simulator class has the classic ABM-style incremental time progression where the clock is advanced with a step size of 1, and for each time step, `model.step` gets called. The DEVS simulator class is a full-blown DES event scheduling simulator with float-based time units, so the clock is advanced based on the timestamp of the next event (resulting in variable time steps). 

I personally would prefer MESA to include event scheduling by default. Also, as discussed in the developer meeting, this adds very little additional complexity while exposing the full power of event scheduling to the user. So why not include it by default in MESA? 

The ABMSimulator class in my code is the most basic way of offering a hybrid to users coming from ABM models because it sticks with discrete time steps and is still centered on `model.step` / `go` in NetLogo, etc.  Since it is trivial to add a DEVS simulator because all the pieces are there already, I just included it. 

On the goal, I am not sure what one would want to add beyond the DEVS Simulator. Yes, you could add a flow library with entities, queues, and processes. This would mimic, _e.g._, the Arena Discrete Event Simulation package. However, once you shift to a more object-oriented programming style for DES, stuff like a flow library doesn't make as much sense anymore.  For reference, I learned DES through Arena, and worked with a whole range of other DES software packages (e.g., eM-plant/Plant simulation, enterprise dynamics, Simio). Once I started using OO with direct event scheduling, I never really looked back because, for me, it was the most expressive way of making these kinds of simulations. Happy to talk more if it is useful.

**Reply by quaquel** [reply]
I started reading the paper you provided. I don't particularly agree with their characterization of DES. It seems to target what I called a "flow library" while ignoring the event scheduling formalism which I used here.

**Reply by rht** [reply]
In what way is a flow library equipped with entities, etc, less expressive than an OOP discrete event framework? Is there a trade-off between expressiveness and performance? Is it similar to the situation with the current data collector being both not general and slow, and the new in-development API that is strictly better?

**Reply by rht** [reply]
For those with no institutional access to the MASON paper, here is a [public link](https://cs.gmu.edu/~sean/papers/annsim23des.pdf) to it (or there is always Nexus [Standard Template Construct](https://www.reddit.com/r/scihub/comments/12detqs/standard_template_construct_store_and_search_the/), since Sci-Hub may be up, but has been frozen since 2020).

**Reply by quaquel** [reply]
You can think of a flow library as a domain-specific language. In a flow library, entities arrive in the system, are sent to queues, claim a resource/server for a specific time, release the resource, and flow to the next queue and server combination. This way of representing systems makes perfect sense for, _e.g._, modeling call centers or logistic systems. Typically, a flow library is also built on top of an event list, as presented here.

However, imagine building a microscopic discrete event simulation of cars on a highway or trams driving around a city. Here, you have objects (cars/trams) reacting to their environment and responding to this by accelerating and braking. All this is still easily expressible through event scheduling, but it makes no sense to express this through entities, queues, and processes. It can be done, as I know from personal experience, but it is torture. 

Here's a nice historical fact: The first Object-oriented programming languages were discrete event simulation packages. OO emerged because of the "naturalness" of describing real-world phenomena for discrete event simulation. It was only later that computer scientists started to adopt OO ideas into the design of programming languages. 

On the MASON paper, I read it yesterday. They talk about DES, but they really mean what I call a flow library. For example, they claim that DES cannot be used to model social phenomena. This is true for a flow library but is demonstrably not true for event scheduling in general.

**Reply by rht** [reply]
> Here's a nice historical fact: The first Object-oriented programming languages were discrete event simulation packages. OO emerged because of the "naturalness" of describing real-world phenomena for discrete event simulation. It was only later that computer scientists started to adopt OO ideas into the design of programming languages.

That's even earlier than the earliest incarnation of modern OO paradigm in a programming language that I know of, Smalltalk (late 1960s), which views objects as tiny independent "computers" that interact with each other via message passing. [Wikipedia](https://en.wikipedia.org/wiki/Object-oriented_programming) said late 1950's was the earliest emergence the concept among the computer scientists.

Though to me, DEVS is more naturally described with message passing (which could be bidirectional) than pub-sub (which is a broadcast, where the sender needs not to be aware of the receivers), even though both are decentralized systems. I suppose the flow library design would be a centralized system?

An incarnation of message passing OOP in Python would be the object method invocation ([Wikipedia article on OOP](https://en.wikipedia.org/wiki/Object-oriented_programming#Dynamic_dispatch/message_passing)). But NetLogo's OOP syntax seems to be more explicit in conveying that it is a "message passing".

**Reply by quaquel** [reply]
My understanding is that discrete event simulation started with [Simscript](https://www.rand.org/content/dam/rand/pubs/research_memoranda/2009/RM3310.pdf) developed at the RAND corporation. Here you already see the use of an EventList, and entities with attributes. [Simulate](https://en.wikipedia.org/wiki/Simula) then builds on this and is considered the first object oriented language. Version I (1962) was targeted at simulations (like Simscript), vesrion II (1967) was a general purpose programming language. Simula had a great influence on the developers of both C++ and Java. Smalltalk actually appeared only in 1972, so a bit later.

**Comment by rht** [comment]
> While the regrowth of the grass patches is handled through event scheduling.

Regarding with performance, I know it's less efficient with stepping all the GrassPatch's at each step, but I reformulated the step method as
```python
def step(self):
    self.height = min(self.height + 1, self.model.grass_regrowth_time)
    self.fully_grown = self.height >= self.max_height
```
This version is about as concise as the DEVS version, while at the same time it can be vectorized, and hence can be a property layer bulk operation, instead of agent steps, which is fast as well. It makes me wonder if the framework language is not Python, but instead is in Rust, which one would be faster.

But this is a very specific example where the grass FSM dynamics is independent of its surroundings. If the grass patch growth rate depends on whether the surrounding is arid/fertile, then a property layer solution might be faster, because dependence on the nearest neighbor can be viewed as a geographical property instead of time-based property.

**Reply by quaquel** [reply]
I agree that for the WolfSheep model, implementing grass as a property layer is sensible.

However, the purpose here is to showcase event scheduling and how it can be combined with normal ABM-style scheduling. The purpose here, explicitly, is not to find the fastest implementation. For the fastest implementation, we can take a look once the DiscreteSpaces stuff has been merged and property grids are made to work within it. The fastest implementation is relevant for the benchmarks with other libraries.

Also, I added a second example using [Epstein's civil violence model](https://github.com/quaquel/mesa/blob/devs/mesa/experimental/devs/examples/epstein_civil_violence.py).

**Comment by rht** [comment]
The [EventList](https://github.com/quaquel/mesa/blob/b7457c551b0ce175f0c7ca34a67922003314baa4/mesa/experimental/devs/eventlist.py#L78)'s concrete implementation is based on [EventListHeap](https://github.com/averbraeck/pydsol-core/blob/c34ba3b9d3b16e0c3f52e0dbe4fb7a858020f74c/src/pydsol/core/eventlist.py#L88) in pydsol. a DEVS is more efficient than a simulation in ABM format if the number of events << number of ticks in ABM. If they are of the same order of magnitude, then the push operations on the heap tree become expensive.

For a hybrid framework, shouldn't the model step be separated from the EventList, because it happens at every tick anyway? The EventList may be reserved for rare events, and the current pop() of the heap shouldn't be cleared unless it is at the right tick.

**Reply by quaquel** [reply]
The developer of pydsol did extensive performance testing. The overhead should be really small, but it is worth testing.

The problem with separating `model.step` from the event list is that you must do additional bookkeeping. This implementation has a single source for truth about the simulation time.

**Reply by rht** [reply]
> For a hybrid framework, shouldn't the model step be separated from the EventList, because it happens at every tick anyway? The EventList may be reserved for rare events, and the current pop() of the heap shouldn't be cleared unless it is at the right tick.

Never mind, this proposal is probably not faster than a heap push, because pushing to a heap is ensuring that the rare event happens at the right tick, and so is almost the same as having them separate, because the checking at each tick is almost the same as a single push operation on the heap that includes the ABM steps.

**Reply by rht** [reply]
I think the `__cmp__` method of the SimEvent can be significantly sped up in Rust, if the EventList heap data storage is in Rust as well. Python comparator is usually slow.

**Reply by quaquel** [reply]
yes, if we go done the Rust route and integrate some kind of DEVS into MESA, it would be logical to port all this to Rust.
